File: Test/TestCases/node-addon-api/object/object.cs
using System.Runtime.CompilerServices;
using NodeApi;

namespace NodeApiTest;

public partial class TestObject : TestHelper, ITestObject
{
    private static bool s_testValue = true;

    // Used to test void* Data() integrity
    private class UserDataHolder
    {
        public int Value { get; set; }
    }

    private static JSValue TestGetter(JSCallbackArgs args)
    {
        return JSValue.GetBoolean(s_testValue);
    }

    private static JSValue TestSetter(JSCallbackArgs args)
    {
        s_testValue = (bool)args[0];
        return JSValue.Undefined;
    }

    private static JSValue TestGetterWithUserData(JSCallbackArgs args)
    {
        if (args.Data is UserDataHolder holder)
        {
            return (JSValue)holder.Value;
        }
        return JSValue.Undefined;
    }

    private static JSValue TestSetterWithUserData(JSCallbackArgs args)
    {
        if (args.Data is UserDataHolder holder)
        {
            holder.Value = (int)args[0];
        }
        return JSValue.Undefined;
    }

    private static JSValue TestFunction(JSCallbackArgs args)
    {
        return true;
    }

    private static JSValue TestFunctionWithUserData(JSCallbackArgs args)
    {
        if (args.Data is UserDataHolder holder)
        {
            return holder.Value;
        }
        return JSValue.Undefined;
    }

    private static JSValue GetPropertyNames(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        return obj.GetPropertyNames();
    }

    private static JSValue DefineProperties(JSCallbackArgs args)
    {
        JSValue obj = args[0];

        JSValue trueValue = JSValue.True;
        UserDataHolder holder = new()
        {
            Value = 1234
        };

        obj.DefineProperties(
            JSPropertyDescriptor.Accessor("readonlyAccessor", TestGetter),
            JSPropertyDescriptor.Accessor("readWriteAccessor", TestGetter, TestSetter),
            JSPropertyDescriptor.Accessor("readonlyAccessorWithUserData",
                                          TestGetterWithUserData,
                                          data: holder),
            JSPropertyDescriptor.Accessor("readWriteAccessorWithUserData",
                                          TestGetterWithUserData,
                                          TestSetterWithUserData,
                                          data: holder),
            JSPropertyDescriptor.ForValue("readonlyValue", trueValue),
            JSPropertyDescriptor.ForValue("readWriteValue", trueValue, JSPropertyAttributes.Writable),
            JSPropertyDescriptor.ForValue("enumerableValue", trueValue, JSPropertyAttributes.Enumerable),
            JSPropertyDescriptor.ForValue("configurableValue", trueValue, JSPropertyAttributes.Configurable),
            JSPropertyDescriptor.Function("function", TestFunction),
            JSPropertyDescriptor.Function("functionWithUserData",
                                          TestFunctionWithUserData,
                                          data: holder)
        );
        return JSValue.Undefined;
    }

    private static JSValue DefineValueProperty(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        JSValue name = args[1];
        JSValue value = args[2];
        obj.DefineProperties(JSPropertyDescriptor.ForValue(name, value));
        return JSValue.Undefined;
    }

    private static JSValue CreateObjectUsingMagic(JSCallbackArgs args)
    {
        JSValue obj = JSValue.CreateObject();
        obj["cp_false"] = false;
        obj["cp_true"] = true;
        obj["s_true"] = true;
        obj["s_false"] = false;
        obj["0"] = 0;
        obj[(uint)42] = 120;
        obj["0.0f"] = 0.0f;
        obj["0.0"] = 0.0;
        obj["-1"] = -1;
        obj["foo2"] = "foo"u8;
        obj["foo4"] = "foo";
        obj["circular"] = obj;
        obj["circular2"] = obj;
        return obj;
    }

    private static JSValue Sum(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        long sum = 0;

        foreach ((JSValue _, JSValue value) in obj.Properties)
        {
            sum += (long)value;
        }

        return sum;
    }

    private static JSValue Increment(JSCallbackArgs args)
    {
        JSValue obj = args[0];

        foreach ((JSValue name, JSValue value) in obj.Properties)
        {
            obj[name] = (long)value + 1;
        }

        return JSValue.Undefined;
    }

    private static JSValue InstanceOf(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        JSValue constructor = args[1];
        return obj.InstanceOf(constructor);
    }

    public static JSObject Init() => new()
    {
        Method(GetPropertyNames),
        Method(DefineProperties),
        Method(DefineValueProperty),

        Method(GetPropertyWithNapiValue),
        Method(GetPropertyWithNapiWrapperValue),
        Method(GetPropertyWithLatin1StyleString),
        Method(GetPropertyWithUtf8StyleString),
        Method(GetPropertyWithCSharpStyleString),
        Method(GetPropertyWithUInt32),

        Method(SetPropertyWithNapiValue),
        Method(SetPropertyWithNapiWrapperValue),
        Method(SetPropertyWithLatin1StyleString),
        Method(SetPropertyWithUtf8StyleString),
        Method(SetPropertyWithCSharpStyleString),
        Method(SetPropertyWithUInt32),

        Method(DeletePropertyWithNapiValue),
        Method(DeletePropertyWithNapiWrapperValue),
        Method(DeletePropertyWithLatin1StyleString),
        Method(DeletePropertyWithUtf8StyleString),
        Method(DeletePropertyWithCSharpStyleString),
        Method(DeletePropertyWithUInt32),

        Method(HasOwnPropertyWithNapiValue),
        Method(HasOwnPropertyWithNapiWrapperValue),
        Method(HasOwnPropertyWithLatin1StyleString),
        Method(HasOwnPropertyWithUtf8StyleString),
        Method(HasOwnPropertyWithCSharpStyleString),

        Method(HasPropertyWithNapiValue),
        Method(HasPropertyWithNapiWrapperValue),
        Method(HasPropertyWithLatin1StyleString),
        Method(HasPropertyWithUtf8StyleString),
        Method(HasPropertyWithCSharpStyleString),
        Method(HasPropertyWithUInt32),

        Method(CreateObjectUsingMagic),
        Method(Sum),
        Method(Increment),

        Method(AddFinalizer),

        Method(InstanceOf),

        Method(SubscriptGetWithLatin1StyleString),
        Method(SubscriptGetWithUtf8StyleString),
        Method(SubscriptGetWithCSharpStyleString),
        Method(SubscriptGetAtIndex),
        Method(SubscriptSetWithLatin1StyleString),
        Method(SubscriptSetWithUtf8StyleString),
        Method(SubscriptSetWithCSharpStyleString),
        Method(SubscriptSetAtIndex),
    };
}

File: Test/TestCases/node-addon-api/object/object.js
'use strict';

const assert = require('assert');

module.exports = require('../../common').runTest(test);

function test(binding) {

  function assertPropertyIs(obj, key, attribute) {
    const propDesc = Object.getOwnPropertyDescriptor(obj, key);
    assert.ok(propDesc);
    assert.ok(propDesc[attribute]);
  }

  function assertPropertyIsNot(obj, key, attribute) {
    const propDesc = Object.getOwnPropertyDescriptor(obj, key);
    assert.ok(propDesc);
    assert.ok(!propDesc[attribute]);
  }

  function testDefineProperties() {
    const obj = {};
    binding.object.defineProperties(obj);

    // accessors
    assertPropertyIsNot(obj, 'readonlyAccessor', 'enumerable');
    assertPropertyIsNot(obj, 'readonlyAccessor', 'configurable');
    assert.strictEqual(obj.readonlyAccessor, true);

    assertPropertyIsNot(obj, 'readonlyAccessorWithUserData', 'enumerable');
    assertPropertyIsNot(obj, 'readonlyAccessorWithUserData', 'configurable');
    assert.strictEqual(obj.readonlyAccessorWithUserData, 1234);

    assertPropertyIsNot(obj, 'readWriteAccessor', 'enumerable');
    assertPropertyIsNot(obj, 'readWriteAccessor', 'configurable');
    obj.readWriteAccessor = false;
    assert.strictEqual(obj.readWriteAccessor, false);
    obj.readWriteAccessor = true;
    assert.strictEqual(obj.readWriteAccessor, true);

    assertPropertyIsNot(obj, 'readWriteAccessorWithUserData', 'enumerable');
    assertPropertyIsNot(obj, 'readWriteAccessorWithUserData', 'configurable');
    obj.readWriteAccessorWithUserData = 2;
    assert.strictEqual(obj.readWriteAccessorWithUserData, 2);
    obj.readWriteAccessorWithUserData = -14;
    assert.strictEqual(obj.readWriteAccessorWithUserData, -14);

    // values
    assertPropertyIsNot(obj, 'readonlyValue', 'writable');
    assertPropertyIsNot(obj, 'readonlyValue', 'enumerable');
    assertPropertyIsNot(obj, 'readonlyValue', 'configurable');
    assert.strictEqual(obj.readonlyValue, true);

    assertPropertyIs(obj, 'readWriteValue', 'writable');
    assertPropertyIsNot(obj, 'readWriteValue', 'enumerable');
    assertPropertyIsNot(obj, 'readWriteValue', 'configurable');
    obj.readWriteValue = false;
    assert.strictEqual(obj.readWriteValue, false);
    obj.readWriteValue = true;
    assert.strictEqual(obj.readWriteValue, true);

    assertPropertyIsNot(obj, 'enumerableValue', 'writable');
    assertPropertyIs(obj, 'enumerableValue', 'enumerable');
    assertPropertyIsNot(obj, 'enumerableValue', 'configurable');

    assertPropertyIsNot(obj, 'configurableValue', 'writable');
    assertPropertyIsNot(obj, 'configurableValue', 'enumerable');
    assertPropertyIs(obj, 'configurableValue', 'configurable');

    // functions
    assertPropertyIsNot(obj, 'function', 'writable');
    assertPropertyIsNot(obj, 'function', 'enumerable');
    assertPropertyIsNot(obj, 'function', 'configurable');
    assert.strictEqual(obj.function(), true);
    assert.strictEqual(obj.functionWithUserData(), obj.readonlyAccessorWithUserData);
  }

  testDefineProperties();

  {
    const obj = {};
    const testSym = Symbol('testSym');
    binding.object.defineValueProperty(obj, testSym, 1);
    assert.strictEqual(obj[testSym], 1);
  }

  {
    const testSym = Symbol('testSym');
    const obj = { one: 1, two: 2, three: 3, [testSym]: 4 };
    const arr = binding.object.getPropertyNames(obj);
    assert.deepStrictEqual(arr, ['one', 'two', 'three']);
  }

  {
    const magicObject = binding.object.createObjectUsingMagic();
    assert.deepStrictEqual(magicObject, {
      0: 0,
      42: 120,
      cp_false: false,
      cp_true: true,
      s_true: true,
      s_false: false,
      '0.0f': 0,
      '0.0': 0,
      '-1': -1,
      foo2: 'foo',
      foo4: 'foo',
      circular: magicObject,
      circular2: magicObject
    });
  }

  {
    function Ctor() { }

    assert.strictEqual(binding.object.instanceOf(new Ctor(), Ctor), true);
    assert.strictEqual(binding.object.instanceOf(new Ctor(), Object), true);
    assert.strictEqual(binding.object.instanceOf({}, Ctor), false);
    assert.strictEqual(binding.object.instanceOf(null, Ctor), false);
  }

  if ('sum' in binding.object) {
    {
      const obj = {
        '-forbid': -0x4B1D,
        '-feedcode': -0xFEEDC0DE,
        '+office': +0x0FF1CE,
        '+forbid': +0x4B1D,
        '+deadbeef': +0xDEADBEEF,
        '+feedcode': +0xFEEDC0DE
      };

      let sum = 0;
      for (const key in obj) {
        sum += obj[key];
      }

      assert.strictEqual(binding.object.sum(obj), sum);
    }

    {
      const obj = new Proxy({
        '-forbid': -0x4B1D,
        '-feedcode': -0xFEEDC0DE,
        '+office': +0x0FF1CE,
        '+forbid': +0x4B1D,
        '+deadbeef': +0xDEADBEEF,
        '+feedcode': +0xFEEDC0DE
      }, {
        getOwnPropertyDescriptor(target, p) {
          throw new Error('getOwnPropertyDescriptor error');
        },
        ownKeys(target) {
          throw new Error('ownKeys error');
        }
      });

      assert.throws(() => {
        binding.object.sum(obj);
      }, /ownKeys error/);
    }
  }

  if ('increment' in binding.object) {
    const obj = {
      a: 0,
      b: 1,
      c: 2
    };

    binding.object.increment(obj);

    assert.deepStrictEqual(obj, {
      a: 1,
      b: 2,
      c: 3
    });
  }
}
